# Quandoo Data Analysis Documentation

## 1.Introduction

### Purpose of the Analysis

The purpose of this analysis is to gain insights into the data provided by Quandoo, specifically the reservations.csv and merchants.csv files. The analysis aims to identify the most important cities for the company and provide a summary of key observations along with recommended metrics for business evaluation.

### Data Sources

1. reservations.csv
2. merchants.csv

## 2. Setup and Tools

The analysis was conducted using Jupyter Notebook and PostgreSQL. 

In [2]:
import pandas as pd
import numpy as np
import psycopg2
from sqlalchemy import create_engine, text
import datetime
from IPython.display import HTML

## 3. Importing Data

In [3]:
marchants = pd.read_csv(r"C:\Users\mmoki\OneDrive\Desktop\Quandoo\merchants.csv")
reservations = pd.read_csv(r"C:\Users\mmoki\OneDrive\Desktop\Quandoo\reservations.csv", parse_dates =['reservation_created_date', 'reservation_completed_date'])

## 4. Data Exploration

Conduct initial data exploration to understand the structure and content of the data.

In [5]:
marchants 

merchant_id            city         cuisine
0      260b-bc5c9-f3ea-aedd        Canberra      Australian
1      320b-b3412-aacf-b0fa            Bern         Swedish
2      4aea-a5a3d-3ef3-8b59      Canterbury        High Tea
3      2942-22d35-f1f1-9d85         Ashford   Cake & Coffee
4      0b0d-df1fc-811a-8f6a            Wien         Swedish
...                     ...             ...             ...
32847  a150-059b1-d6af-8411      Düsseldorf  Middle Eastern
32848  7d01-18b57-f142-b902  Recklinghausen  Middle Eastern
32849  1fbe-eaf86-f1d3-802a            Köln  Middle Eastern
32850  cfa0-087ff-644d-9ccb         Krefeld  Middle Eastern
32851  fabd-df7a0-1071-b8a4      Düsseldorf  Middle Eastern

[32852 rows x 3 columns]

In [6]:
reservations

reservation_id  reservation_created_date  \
0        b4ea-a5cb0-10e4-90a4 2023-06-22 10:08:19+00:00   
1        14d4-4e910-1144-a32a 2023-06-22 21:31:03+00:00   
2        c837-79fa0-10ce-a727 2023-06-22 07:31:18+00:00   
3        b362-28100-10e0-a21c 2023-06-22 09:39:37+00:00   
4        d61c-ca6e0-114d-aa03 2023-06-22 22:40:52+00:00   
...                       ...                       ...   
1612125  c6b3-35010-0cdd-9c28 2023-06-17 07:08:37+00:00   
1612126  bda2-23a00-0d59-b5c6 2023-06-17 21:55:59+00:00   
1612127  a750-03aa0-0d03-a803 2023-06-17 11:39:45+00:00   
1612128  e374-4762c-1f7a-913e 2023-06-17 09:13:38+00:00   
1612129  8087-7d0f9-60aa-9d0b 2023-06-17 10:33:38+00:00   

        reservation_completed_date reservation_status           customer_id  \
0        2023-06-24 06:45:00+00:00         successful  2cbc-c40df-c6f4-b9d6   
1        2023-06-23 09:30:00+00:00         successful  1775-52029-4195-b8ae   
2        2023-06-24 05:45:00+00:00           canceled  8968-850e2-844f-811b   
3        2023-06-24 07:00:00+00:00         successful  1b97-7d46a-bef4-a20e   
4        2023-06-23 00:30:00+00:00         successful  9be2-22a6a-dcd5-aa37   
...                            ...                ...                   ...   
1612125  2023-06-18 11:00:00+00:00           canceled  3aad-dc6d4-0d10-968a   
1612126  2023-06-24 16:30:00+00:00           canceled  8151-18a7e-ba90-88bc   
1612127  2023-06-17 16:30:00+00:00           canceled  dcb1-15e26-fc06-af08   
1612128  2023-06-17 10:30:00+00:00           canceled  6997-7f88e-145f-91e1   
1612129  2023-06-17 12:00:00+00:00           canceled  6997-7f88e-145f-91e1   

                  merchant_id    revenue marketing_channel country  
0        511d-d4351-7704-a5b1   4.219409         Affiliate      NZ  
1        f4a5-5c44c-611b-8d48  11.251758            Direct      NZ  
2        f4a5-5c44c-611b-8d48   5.625879          Referral      NZ  
3        f4a5-5c44c-611b-8d48   4.219409             Other      NZ  
4        4bea-ab9e0-ad2d-8075   7.876231             Other      NZ  
...                       ...        ...               ...     ...  
1612125  35ed-de74d-e48f-9d76   1.797591            Direct      TR  
1612126  6aa6-69097-702c-9ad7   1.797591          Referral      TR  
1612127  62a3-3810d-3481-b25c   2.696387         Affiliate      TR  
1612128  b287-76c62-5c56-bd29   1.887471            Direct      TR  
1612129  b287-76c62-5c56-bd29   1.887471            Direct      TR  

[1612130 rows x 9 columns]

## 5. Connect to PostgreSQL

In [8]:
db_params = {
    'host': 'localhost',
    'database': 'Quandoo',
    'user': 'postgres',
    'password': '2168165567'
}

In [10]:
db_params['database'] = 'quandoo'
engine = create_engine(f'postgresql://{db_params["user"]}:{db_params["password"]}@{db_params["host"]}/{db_params["database"]}')

## Define the file path for your CSV files
csv_files = {
    'merchants': r"C:\Users\mmoki\OneDrive\Desktop\Quandoo\merchants.csv",
    'reservations': r"C:\Users\mmoki\OneDrive\Desktop\Quandoo\cleaned_reservations.csv"
}

In [11]:
# Loop through the CSV files and import them into PostgreSQL
for table_name, file_path in csv_files.items():
    df = pd.read_csv(file_path)
    df.to_sql(table_name, engine, if_exists='replace', index=False)

Now that we are done with connecting jupyter notebook and postgeSQL, lets do some more EDA.

In [13]:
%load_ext sql

%sql postgresql://postgres:2168165567@localhost:5432/quandoo

In [14]:
%%sql

SELECT * 
FROM merchants
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/quandoo
10 rows affected.


[('260b-bc5c9-f3ea-aedd', 'Canberra', 'Australian'),
 ('320b-b3412-aacf-b0fa', 'Bern', 'Swedish'),
 ('4aea-a5a3d-3ef3-8b59', 'Canterbury', 'High Tea'),
 ('2942-22d35-f1f1-9d85', 'Ashford', 'Cake & Coffee'),
 ('0b0d-df1fc-811a-8f6a', 'Wien', 'Swedish'),
 ('bf03-3de03-7b46-bd97', 'Wien', 'Cake & Coffee'),
 ('418f-fc212-a6bf-a33f', 'Wien', 'Contemporary'),
 ('ed13-31d9b-b453-bd1c', 'Wien', 'Arabic'),
 ('b219-9937f-25b0-8b2a', 'Wien', 'Contemporary'),
 ('7b52-24c80-da28-a185', 'Wien', 'Latin American')]

In [15]:
%%sql
SELECT * 
FROM reservations
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/quandoo
10 rows affected.


[('b4ea-a5cb0-10e4-90a4', '2023-06-22 10:08:19+00:00', '2023-06-24 06:45:00+00:00', 'successful', '2cbc-c40df-c6f4-b9d6', '511d-d4351-7704-a5b1', 4.219409282700422, 'Affiliate', 'NZ'),
 ('14d4-4e910-1144-a32a', '2023-06-22 21:31:03+00:00', '2023-06-23 09:30:00+00:00', 'successful', '1775-52029-4195-b8ae', 'f4a5-5c44c-611b-8d48', 11.251758087201123, 'Direct', 'NZ'),
 ('c837-79fa0-10ce-a727', '2023-06-22 07:31:18+00:00', '2023-06-24 05:45:00+00:00', 'canceled', '8968-850e2-844f-811b', 'f4a5-5c44c-611b-8d48', 5.625879043600563, 'Referral', 'NZ'),
 ('b362-28100-10e0-a21c', '2023-06-22 09:39:37+00:00', '2023-06-24 07:00:00+00:00', 'successful', '1b97-7d46a-bef4-a20e', 'f4a5-5c44c-611b-8d48', 4.219409282700422, 'Other', 'NZ'),
 ('d61c-ca6e0-114d-aa03', '2023-06-22 22:40:52+00:00', '2023-06-23 00:30:00+00:00', 'successful', '9be2-22a6a-dcd5-aa37', '4bea-ab9e0-ad2d-8075', 7.876230661040787, 'Other', 'NZ'),
 ('d691-10840-10b8-9440', '2023-06-22 04:54:16+00:00', '2023-07-06 05:30:00+00:00', 'successful', '2e61-168ac-4088-9b9e', '639e-ed6f3-8a07-b785', 13.719341471609363, 'Other', 'NZ'),
 ('1a4a-a6ad0-1547-8683', '2023-06-28 00:02:45+00:00', '2023-07-01 06:30:00+00:00', 'successful', '7a44-40674-6fed-a874', 'f4a5-5c44c-611b-8d48', 5.599731212901781, 'Direct', 'NZ'),
 ('23e0-0b760-1584-b8b2', '2023-06-28 07:19:39+00:00', '2023-06-28 07:45:00+00:00', 'successful', '3d37-7032a-9252-a83c', '818d-df08a-f0f7-a70b', 2.8129395218002813, 'Referral', 'NZ'),
 ('2bba-aef20-158c-a45b', '2023-06-28 08:17:07+00:00', '2023-06-30 08:00:00+00:00', 'successful', '7776-69490-0e5f-bf6d', '3909-9b53d-319c-8f3a', 8.438818565400846, 'Other', 'NZ'),
 ('9496-6eda0-1609-961d', '2023-06-28 23:14:52+00:00', '2023-06-29 06:00:00+00:00', 'successful', 'db80-0551e-e322-8033', '8792-2e43b-123d-b9fe', 13.783403656821378, 'Referral', 'NZ')]

In [31]:
%%sql

SELECT 
  COUNT(reservation_id) AS total_reservations
FROM reservations;

 * postgresql://postgres:***@localhost:5432/quandoo
1 rows affected.


[(1612130,)]

In [16]:
%%sql

SELECT
  reservation_status,
  COUNT(reservation_status) AS res_count
FROM
  reservations
GROUP BY
  1
ORDER BY
  2 desc;

 * postgresql://postgres:***@localhost:5432/quandoo
4 rows affected.


[('successful', 1258784),
 ('canceled', 315299),
 ('noshow', 36345),
 ('notified', 1702)]

Analyzing reservation status in this way helps in understanding the distribution of reservations and provides insights into the success rates, cancellation patterns, and customer behavior.

Examining the reservation table, I observe the presence of the marketing_channel field, which I consider significant as it can provide insights into the ways our customers discover our services

In [17]:
%%sql

SELECT
  marketing_channel,
  COUNT(marketing_channel) AS channel_count
FROM
  reservations
GROUP BY
  1
ORDER BY
  2 desc;

 * postgresql://postgres:***@localhost:5432/quandoo
9 rows affected.


[('SEM', 541596),
 ('Referral', 340448),
 ('Direct', 280177),
 ('Other', 242403),
 ('SEO', 124194),
 ('Affiliate', 58823),
 ('CRM', 24275),
 ('Paid Social', 203),
 ('Organic Social', 11)]

These findings provide valuable guidance for shaping the marketing strategy, shedding light on how different channels contribute to reservation generation. SEM stands out as the most effective channel, closely followed by Referral. However, there is room for improvement in the Organic Social channel. Exploring innovative content strategies might be a promising avenue for enhancement.

## 6. Question 1 : Identifying the Most Important Cities

In [20]:
%%sql

SELECT
  m.city,
  ROUND(SUM(r.revenue)::numeric, 2) AS total_revenue
FROM
  merchants m
JOIN
  reservations r
ON
  m.merchant_id = r.merchant_id
GROUP BY
  m.city
ORDER BY
  total_revenue DESC
LIMIT
  5;

 * postgresql://postgres:***@localhost:5432/quandoo
5 rows affected.


[('Singapore', Decimal('1365393.20')),
 ('Wien', Decimal('347852.70')),
 ('Berlin', Decimal('338239.80')),
 ('Melbourne', Decimal('331125.56')),
 ('London', Decimal('303714.16'))]

The analysis of revenue data underscores Singapore's remarkable lead as the top-performing country, demonstrating a significant revenue gap from the next best; therefore, it suggests the need to investigate and potentially implement successful strategies from Singapore in other cities to optimize overall performance.

When dealing with revenue it is important to find what is the difference on monthly based on the requirements of the company.

## 7. Question 2: Summarizing Important Observations

### Key Observations

The merchants file, it consists of three columns: merchant_id, city, and cuisine. The dataset encompasses 32,852 distinct merchant_ids across 1,409 cities and 129 cuisines. While there are null values present in the cuisine column, for the current analysis, we will leave them unaddressed.

The reservation file is a key focus for our analysis. It holds important information about when each reservation is created and completed, helping us track the timeline of each booking. Additionally, the merchant_id in this file allows us to connect and combine data with merchant table. The revenue details in the reservation file are crucial for understanding the financial impact of our reservations, especially when looking at contributions from different countries. This helps us make informed decisions about resource allocation. Furthermore, the reservation file includes information about the marketing channel used for each reservation. This insight is valuable for assessing the effectiveness of our marketing strategies and identifying the channels that attract customers.

In essence, the reservation file is a central piece in our analysis, providing insights into timing, relationships, revenue, and marketing efforts that collectively contribute to a better understanding of our business operations

### Important Metrics for Quandoo

1. Total Revenue - Understanding the overall revenue generated is crucial for evaluating business performance.

In [21]:
%%sql

SELECT
ROUND(SUM(revenue)::numeric, 2) AS total_revenue 
FROM reservations

 * postgresql://postgres:***@localhost:5432/quandoo
1 rows affected.


[(Decimal('6876639.36'),)]

2. Month to Month Revenue and Percentage Difference - Monitoring monthly revenue trends and percenatge difference aids in identifying growth or decline patterns. 

In [22]:
%%sql

WITH
  monthly_revenue AS (
  SELECT
    DATE_TRUNC('month', CAST(reservation_created_date AS TIMESTAMP)) AS month,
    country,
    SUM(revenue) OVER (PARTITION BY country, DATE_TRUNC('month', CAST(reservation_created_date AS TIMESTAMP))
    ORDER BY
      reservation_created_date) AS running_total_revenue
  FROM
    reservations )
SELECT
  month,
  country,
  CAST(MAX(running_total_revenue) AS numeric(10,
      2)) AS monthly_running_total_revenue,
  CAST( (MAX(running_total_revenue) - LAG(MAX(running_total_revenue)) OVER (PARTITION BY country ORDER BY month)) / NULLIF(LAG(MAX(running_total_revenue)) OVER (PARTITION BY country ORDER BY month), 0) * 100 AS numeric(10,
      2) ) AS percentage_change
FROM
  monthly_revenue
GROUP BY
  1,
  2
ORDER BY
  2,
  1;

 * postgresql://postgres:***@localhost:5432/quandoo
120 rows affected.


[(datetime.datetime(2023, 1, 1, 0, 0), 'AT', Decimal('74035.80'), None),
 (datetime.datetime(2023, 2, 1, 0, 0), 'AT', Decimal('62873.00'), Decimal('-15.08')),
 (datetime.datetime(2023, 3, 1, 0, 0), 'AT', Decimal('65411.50'), Decimal('4.04')),
 (datetime.datetime(2023, 4, 1, 0, 0), 'AT', Decimal('64314.65'), Decimal('-1.68')),
 (datetime.datetime(2023, 5, 1, 0, 0), 'AT', Decimal('54276.65'), Decimal('-15.61')),
 (datetime.datetime(2023, 6, 1, 0, 0), 'AT', Decimal('50979.60'), Decimal('-6.07')),
 (datetime.datetime(2023, 7, 1, 0, 0), 'AT', Decimal('50339.55'), Decimal('-1.26')),
 (datetime.datetime(2023, 8, 1, 0, 0), 'AT', Decimal('43247.75'), Decimal('-14.09')),
 (datetime.datetime(2023, 9, 1, 0, 0), 'AT', Decimal('37477.95'), Decimal('-13.34')),
 (datetime.datetime(2023, 10, 1, 0, 0), 'AT', Decimal('38133.40'), Decimal('1.75')),
 (datetime.datetime(2023, 1, 1, 0, 0), 'AU', Decimal('114034.29'), None),
 (datetime.datetime(2023, 2, 1, 0, 0), 'AU', Decimal('111275.62'), Decimal('-2.42')),
 (datetime.datetime(2023, 3, 1, 0, 0), 'AU', Decimal('135727.45'), Decimal('21.97')),
 (datetime.datetime(2023, 4, 1, 0, 0), 'AU', Decimal('129009.86'), Decimal('-4.95')),
 (datetime.datetime(2023, 5, 1, 0, 0), 'AU', Decimal('98068.51'), Decimal('-23.98')),
 (datetime.datetime(2023, 6, 1, 0, 0), 'AU', Decimal('85020.63'), Decimal('-13.30')),
 (datetime.datetime(2023, 7, 1, 0, 0), 'AU', Decimal('81087.57'), Decimal('-4.63')),
 (datetime.datetime(2023, 8, 1, 0, 0), 'AU', Decimal('85558.85'), Decimal('5.51')),
 (datetime.datetime(2023, 9, 1, 0, 0), 'AU', Decimal('70493.01'), Decimal('-17.61')),
 (datetime.datetime(2023, 10, 1, 0, 0), 'AU', Decimal('59155.75'), Decimal('-16.08')),
 (datetime.datetime(2023, 1, 1, 0, 0), 'CH', Decimal('7246.79'), None),
 (datetime.datetime(2023, 2, 1, 0, 0), 'CH', Decimal('6817.65'), Decimal('-5.92')),
 (datetime.datetime(2023, 3, 1, 0, 0), 'CH', Decimal('6388.74'), Decimal('-6.29')),
 (datetime.datetime(2023, 4, 1, 0, 0), 'CH', Decimal('4215.25'), Decimal('-34.02')),
 (datetime.datetime(2023, 5, 1, 0, 0), 'CH', Decimal('2732.72'), Decimal('-35.17')),
 (datetime.datetime(2023, 6, 1, 0, 0), 'CH', Decimal('2881.16'), Decimal('5.43')),
 (datetime.datetime(2023, 7, 1, 0, 0), 'CH', Decimal('2384.94'), Decimal('-17.22')),
 (datetime.datetime(2023, 8, 1, 0, 0), 'CH', Decimal('2596.92'), Decimal('8.89')),
 (datetime.datetime(2023, 9, 1, 0, 0), 'CH', Decimal('2372.35'), Decimal('-8.65')),
 (datetime.datetime(2023, 10, 1, 0, 0), 'CH', Decimal('2475.74'), Decimal('4.36')),
 (datetime.datetime(2023, 1, 1, 0, 0), 'DE', Decimal('151611.15'), None),
 (datetime.datetime(2023, 2, 1, 0, 0), 'DE', Decimal('157520.00'), Decimal('3.90')),
 (datetime.datetime(2023, 3, 1, 0, 0), 'DE', Decimal('182839.00'), Decimal('16.07')),
 (datetime.datetime(2023, 4, 1, 0, 0), 'DE', Decimal('177620.05'), Decimal('-2.85')),
 (datetime.datetime(2023, 5, 1, 0, 0), 'DE', Decimal('145775.30'), Decimal('-17.93')),
 (datetime.datetime(2023, 6, 1, 0, 0), 'DE', Decimal('120043.60'), Decimal('-17.65')),
 (datetime.datetime(2023, 7, 1, 0, 0), 'DE', Decimal('135181.40'), Decimal('12.61')),
 (datetime.datetime(2023, 8, 1, 0, 0), 'DE', Decimal('136448.60'), Decimal('0.94')),
 (datetime.datetime(2023, 9, 1, 0, 0), 'DE', Decimal('117404.05'), Decimal('-13.96')),
 (datetime.datetime(2023, 10, 1, 0, 0), 'DE', Decimal('132787.35'), Decimal('13.10')),
 (datetime.datetime(2023, 1, 1, 0, 0), 'FI', Decimal('28189.48'), None),
 (datetime.datetime(2023, 2, 1, 0, 0), 'FI', Decimal('36462.56'), Decimal('29.35')),
 (datetime.datetime(2023, 3, 1, 0, 0), 'FI', Decimal('42160.54'), Decimal('15.63')),
 (datetime.datetime(2023, 4, 1, 0, 0), 'FI', Decimal('34840.53'), Decimal('-17.36')),
 (datetime.datetime(2023, 5, 1, 0, 0), 'FI', Decimal('19497.91'), Decimal('-44.04')),
 (datetime.datetime(2023, 6, 1, 0, 0), 'FI', Decimal('14560.65'), Decimal('-25.32')),
 (datetime.datetime(2023, 7, 1, 0, 0), 'FI', Decimal('15425.39'), Decimal('5.94')),
 (datetime.datetime(2023, 8, 1, 0, 0), 'FI', Deci

3. Average Duration per Reservations - Knowing how long customers typically spend during reservations helps in optimizing table turnover and customer satisfaction.

In [24]:
%%sql

SELECT
    country,
    ROUND(AVG(EXTRACT(EPOCH FROM (reservation_completed_date::timestamp - reservation_created_date::timestamp)) / 86400)::numeric, 2) AS average_duration
FROM
    reservations
WHERE
    reservation_status = 'successful'
GROUP BY
    country
ORDER BY
    average_duration DESC;

 * postgresql://postgres:***@localhost:5432/quandoo
12 rows affected.


[('GB', Decimal('6.70')),
 ('CH', Decimal('6.17')),
 ('TR', Decimal('6.01')),
 ('FI', Decimal('5.63')),
 ('AT', Decimal('5.48')),
 ('DE', Decimal('5.41')),
 ('IT', Decimal('5.22')),
 ('HK', Decimal('4.68')),
 ('NZ', Decimal('3.68')),
 ('NL', Decimal('3.61')),
 ('SG', Decimal('3.52')),
 ('AU', Decimal('3.46'))]

Addressing this task proved to be challenging, requiring a revisitation and a deepened understanding of certain concepts, as documented in the PostgreSQL documentation on datetime functions. The objective was to determine the duration it takes for customers to progress from initiating a reservation to successfully completing it. My emphasis was primarily on reservations that culminated in success

Dealing with merchants it might be important to check who are the top ranked merchants

4. Average Revenue per Reservations - Analyzing the average revenue per reservation helps in optimizing pricing and marketing strategies.

In [25]:
%%sql

SELECT
    country,
    ROUND(AVG(revenue)::numeric, 2) AS average_revenue_per_reservation
FROM
    reservations
GROUP BY
    1
ORDER BY
    2 DESC;

 * postgresql://postgres:***@localhost:5432/quandoo
12 rows affected.


[('NZ', Decimal('5.53')),
 ('IT', Decimal('5.22')),
 ('AU', Decimal('5.08')),
 ('GB', Decimal('5.00')),
 ('FI', Decimal('4.95')),
 ('DE', Decimal('4.30')),
 ('SG', Decimal('3.82')),
 ('HK', Decimal('3.19')),
 ('AT', Decimal('3.10')),
 ('NL', Decimal('2.90')),
 ('CH', Decimal('2.86')),
 ('TR', Decimal('1.24'))]

Analyzing the average revenue per reservation provides insights into why some countries have high conversion rates. For instance, Finland stands out with an impressive average of 4.95, indicating successful strategies. This suggests a connection between revenue and successful conversions, urging further exploration into the strategies that contribute to Finland's positive performance

5. Top Merchants - Recognizing top-performing merchants allows focusing on successful partnerships and strategies.

In [26]:
%%sql

WITH RankedMerchants AS (
    SELECT
        m.merchant_id,
        m.city,
	    r.country,
        COUNT(*) AS reservation_count,
        ROW_NUMBER() OVER (ORDER BY COUNT(*) DESC) AS merchant_rank
    FROM
        reservations r
    JOIN
        merchants m ON r.merchant_id = m.merchant_id
    GROUP BY
        1, 2, 3
)
SELECT
    merchant_id,
	city,
	country,
    reservation_count
FROM
    RankedMerchants
WHERE
    merchant_rank <= 10

 * postgresql://postgres:***@localhost:5432/quandoo
10 rows affected.


[('51b1-1ff6a-b6b5-8457', 'Bochum', 'DE', 10182),
 ('94fe-edd11-88f0-bc3c', 'Wien', 'AT', 7879),
 ('ccad-d7685-b36a-9671', 'Worcester', 'GB', 7657),
 ('b7d9-9a7d7-446f-bed6', 'Wien', 'AT', 6973),
 ('e566-680e8-d347-a1d6', 'Gelsenkirchen', 'DE', 6166),
 ('e0e1-1d2c3-d9a0-a41f', 'Singapore', 'SG', 6124),
 ('e9ef-fec50-76b5-bc61', 'Köln', 'DE', 5860),
 ('99ab-bb37d-7bd8-8fe1', 'Frankfurt', 'DE', 5348),
 ('36f0-0eb8a-70b2-9abf', 'Singapore', 'SG', 5343),
 ('1021-1625f-7704-a939', 'Wien', 'AT', 4938)]

6. Top Cuisines per Country - Identifying popular cuisines per country aids in tailoring marketing strategies and improving offerings.

In [30]:
%%sql

WITH RankedCuisines AS (
    SELECT
        r.country,
        m.cuisine,
        COUNT(*) AS cuisine_count,
        ROW_NUMBER() OVER (PARTITION BY r.country ORDER BY COUNT(*) DESC) AS cuisine_rank
    FROM
        reservations r
    JOIN
        merchants m ON r.merchant_id = m.merchant_id
    GROUP BY
        1,2
)
SELECT
    country,
    cuisine,
    cuisine_count
FROM
    RankedCuisines
WHERE
    cuisine_rank = 1
ORDER BY 3 desc;

 * postgresql://postgres:***@localhost:5432/quandoo
12 rows affected.


[('SG', 'Chinese', 85202),
 ('DE', 'Eat & Drink', 83106),
 ('AU', 'Asian', 61338),
 ('IT', 'Italian', 60967),
 ('AT', 'Eat & Drink', 47790),
 ('GB', 'Drinks', 27872),
 ('TR', None, 14963),
 ('FI', 'European', 11238),
 ('NL', 'Seafood', 6382),
 ('CH', 'Middle Eastern', 3894),
 ('HK', 'Asian', 3071),
 ('NZ', 'Eat & Drink', 977)]

7. Conversion rate - Calculating the conversion rate provides insights into the effectiveness of the reservation process.

In [28]:
%%sql

WITH reservation_counts AS (
    SELECT
        country,
        COUNT(DISTINCT reservation_id) AS total_reservations,
        COUNT(DISTINCT CASE WHEN reservation_status = 'successful' THEN reservation_id END) AS successful_reservations
    FROM
        reservations
    GROUP BY
        1
)
SELECT
    country,
    total_reservations,
    successful_reservations,
    ROUND((successful_reservations * 100.0) / total_reservations, 2)::numeric AS conversion_rate
FROM
    reservation_counts
ORDER BY
    4 DESC;

 * postgresql://postgres:***@localhost:5432/quandoo
12 rows affected.


[('FI', 50313, 42942, Decimal('85.35')),
 ('CH', 14049, 11876, Decimal('84.53')),
 ('NL', 24693, 20754, Decimal('84.05')),
 ('AU', 190829, 160145, Decimal('83.92')),
 ('AT', 174413, 141759, Decimal('81.28')),
 ('NZ', 2375, 1895, Decimal('79.79')),
 ('DE', 338867, 263752, Decimal('77.83')),
 ('GB', 165644, 128333, Decimal('77.48')),
 ('IT', 216645, 163962, Decimal('75.68')),
 ('SG', 395358, 296184, Decimal('74.92')),
 ('HK', 9758, 6921, Decimal('70.93')),
 ('TR', 29186, 20261, Decimal('69.42'))]

## 8. Conclusion 

### Summary of Findings 

In conclusion, the analysis of Quandoo's reservation and merchant data has provided valuable insights into various aspects of the business. Here's a summary of the key findings:

### Revenue Insights 

 -  Total Revenue:
The total revenue for the analyzed period stands at 6,876,639.69. This  figure provides a comprehensive view of the company's financial performance.

 - Revenue per City: 
Singapore contributes signficantly to the overall revenue, indicating strategic ares for business focus and expansion.
 
### Customer Behavior 

 - Average Duration per Reservation:
The average duration per successful reservation for each countries shows Austia with least average duration whereas Great Bratain withe highes.

 - Top Cuisines per Country:
Analyzing popular cuisines per country, such as Singapore preferable cusines is Chinese and Germans prefer Eat & Drink, helps tailor marketing strategies and refine offerings to align with customer preferences. 

### Operational Metrics:

 - Number of reservations:
The total number of reservations made during the analyzed period is 1,612,130. With successful reservations as 1,258,784, canceled 315,299,
noshow	36345, and notified	1702.

 - Conversion Rate: 
The conversion rate, calculated per country provides insights into the effectiveness of the reservation process. We have 5 countries achiving conversions rates surpassing the 80% mark. Finland and Switzerland been the top 2.

 - Average Revenue per Reservation:
The average revenue per reservation per countr is a crucial metric for optimizing pricing and marketing strategies. The hightest is from New Zealand 5.53 euros and the least is Turkey 1.24 euros.


### Recommedations:

Based on these findings, severaal recommendations can be considered:

- Focus on High-Performing Cities:
Allocate resources and markeitng effors towards cities with the highest revenue contribution.

- Strengthen Relations with Top Merchants:
Foster strong partnership with top-performing merchants to improve customer satisfaction and loyalty.

- Cuisine Marketing strategies:
Targeted marketing strategies for popular cuisines in each country to attract a wider customer base.

